source/opt/passes.cpp

```diff
@@ -71,5 +79,93 @@ bool FreezeSpecConstantValuePass::Process(ir::Module* module) {
+  // working list and count number of real uses for constants. Uses in
+  // annotation instructions do not count.
+  std::unordered_map<ir::Instruction*, size_t> use_counts;
+  std::vector<ir::Instruction*> constants = module->GetConstants();
+  for (auto* c : constants) {
+    uint32_t const_id = c->result_id();
+    size_t count = 0;
+    if (analysis::UseList* uses = def_use.GetUses(const_id)) {
+      count =
+          std::count_if(uses->begin(), uses->end(), [](const analysis::Use& u) {
+            return !(ir::IsAnnotationInst(u.inst->opcode()) ||
+                     ir::IsDebugInst(u.inst->opcode()));
+          });
+    }
+    use_counts[c] = count;
+    if (!count) {
+      working_list.insert(c);
+    }
+  }
+
+  // Start from the constants with 0 uses, back trace through the def-use chain
+  // to find all dead constants.
+  std::unordered_set<ir::Instruction*> dead_consts;
+  while (!working_list.empty()) {
+    ir::Instruction* inst = *working_list.begin();
+    // Back propagate if the instruction contains IDs in its operands.
+    switch (inst->opcode()) {
+      case SpvOp::SpvOpConstantComposite:
+      case SpvOp::SpvOpSpecConstantComposite:
+      case SpvOp::SpvOpSpecConstantOp:
+        for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+          // SpecConstantOp instruction contains 'opcode' as its operand. Need
+          // to exclude such operands when decreasing uses.
+          if (inst->GetInOperand(i).type != SPV_OPERAND_TYPE_ID) {
+            continue;
+          }
+          uint32_t operand_id = inst->GetSingleWordInOperand(i);
+          ir::Instruction* def_inst = def_use.GetDef(operand_id);
+          // If the use_count does not have any count for the def_inst,
+          // def_inst must not be a constant, and should be ignored here.
+          if (!use_counts.count(def_inst)) {
+            continue;
+          }
+          // The number of uses should never be less then 0, so it can not be
+          // less than 1 before it decreases.
+          assert(use_counts[def_inst] > 0);
+          --use_counts[def_inst];
+          if (!use_counts[def_inst]) {
+            working_list.insert(def_inst);
+          }
+        }
+        break;
+      default:
+        break;
+    }
+    dead_consts.insert(inst);
+    working_list.erase(inst);
+  }
+
+  // Find all annotation and debug instructions that are referencing dead
+  // constants.
+  std::unordered_set<ir::Instruction*> dead_others;
+  for (auto* dc : dead_consts) {
+    if (analysis::UseList* uses = def_use.GetUses(dc->result_id())) {
+      for (const auto& u : *uses) {
+        if (ir::IsAnnotationInst(u.inst->opcode()) ||
+            ir::IsDebugInst(u.inst->opcode())) {
+          dead_others.insert(u.inst);
+        }
+      }
+    }
+  }
+
+  // Turn all dead instructions and uses of them to nop
+  for (auto* dc : dead_consts) {
+    def_use.KillDef(dc->result_id());
+  }
+  for (auto* da : dead_others) {
+    da->ToNop();
```

Comment:
Actually this code snippet has been exercised. The assembly code builder adds `OpDecorate SpecId` instruction for all the primary spec constants, and each defined constant has an `OpName` instruction. Those instructions must be removed after the pass so that the result can match with the expected result, in which we don't add those constants.